File: resources/ios/Sources/Download/DownloadState.swift
import Foundation

final class DownloadState {
    let requestId: String
    let request: URLRequest
    let destination: URL
    let destinationKey: String
    let partial: URL
    let overwrite: Bool
    let policy: RetryPolicy?
    var cancelled = false
    var terminal = false
    var lastProgressAt: TimeInterval = 0
    var attempt = 0
    weak var task: URLSessionTask?
    var retryWorkItem: DispatchWorkItem?

    init(requestId: String, request: URLRequest, destination: URL, destinationKey: String, partial: URL, overwrite: Bool, policy: RetryPolicy?) {
        self.requestId = requestId
        self.request = request
        self.destination = destination
        self.destinationKey = destinationKey
        self.partial = partial
        self.overwrite = overwrite
        self.policy = policy
    }
}

File: resources/android/src/com/victorycodedev/plugins/nativephp_fetch/download/DownloadState.kt
package com.victorycodedev.plugins.nativephp_fetch

import androidx.fragment.app.FragmentActivity
import okhttp3.Call
import java.io.File
import java.io.IOException
import java.util.concurrent.ScheduledFuture

internal class DownloadStartException(val code: String, override val message: String) : Exception(message)
internal class DownloadFileException(val code: String, override val message: String) : IOException(message)

internal data class DownloadState(
    val activity: FragmentActivity,
    val requestId: String,
    val destination: File,
    val destinationKey: String,
    val partial: File,
    val overwrite: Boolean,
    val policy: RetryPolicy?,
    @Volatile var cancelled: Boolean = false,
    @Volatile var terminal: Boolean = false,
    @Volatile var attempt: Int = 0,
    @Volatile var call: Call? = null,
    @Volatile var retryFuture: ScheduledFuture<*>? = null,
)
